populator-machinery/controller.go

```diff
@@ -527,13 +556,49 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
 	}
 
 	// Look for the populator pod
-	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
-	c.addNotification(key, "pod", c.populatorNamespace, podName)
 	var pod *corev1.Pod
-	pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
+	if c.usePod {
+		c.addNotification(key, "pod", c.populatorNamespace, podName)
+		pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	// If populate data without populator pod, and orginal StorageClass's VolumeBindingMode is VolumeBindingWaitForFirstConsumer,
+	// create a StorageClass with VolumeBindingImmediate for pvcPrime
+	storageClassPrimeName := *pvc.Spec.StorageClassName
+	if !c.usePod && storageClass.VolumeBindingMode != nil && storagev1.VolumeBindingWaitForFirstConsumer == *storageClass.VolumeBindingMode {
+		storageClassPrimeName = populatorStorageClassPrefix + "-" + *pvc.Spec.StorageClassName
+		scPrime, err := c.scLister.Get(storageClassPrimeName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			c.addNotification(key, "sc", "", storageClassPrimeName)
+			bm := storagev1.VolumeBindingImmediate
+			scPrime = &storagev1.StorageClass{
+				ObjectMeta:           metav1.ObjectMeta{Name: storageClassPrimeName},
```

Comment:
how is the storage class cleaned up? If we are creating a prime storage class per original storage class cleanup, we should ideally cleanup the prime when the original storage class is deleted (so needs a watch on storage class events and necessary rbacs)

Reply:
That will be out of scope of syncPvc, we need to add a syncStorageClass function

Reply:
Ack. please put a TODO

Reply:
Done